Heterogeneous Explanation

avgupta456 opened the pull request:
Starts heterogeneous explanations using the new `Explainer` framework. Creates `HeterogeneousExplanation` and extends `Explainer`, `ExplainerAlgorithm`, and `DummyExplainer` to handle heterogeneous graphs. Tests and functionalities are partially complete, looking to get some initial validation before continuing.

Todo
- Heterogeneous explanation tests
- Add and test thresholding for heterogeneous explanations
- Finish some heterogeneous explanation methods
- Extend GNNExplainer (maybe in separate PR)

Closes https://github.com/pyg-team/pytorch_geometric/issues/6014

wsad1 commented on torch_geometric/explain/algorithm/utils.py:
```diff
@@ -56,3 +56,12 @@ def clear_masks(model: torch.nn.Module):
     return module
+
+
+def clear_hetero_masks(model: torch.nn.Module):
```
`clear_masks` should work for heterogenous models too. so we don't need this function.

wsad1 commented on torch_geometric/explain/explainer.py:
```diff
@@ -39,11 +44,19 @@ def __init__(
+        # check if model is heterogeneous
+        self.is_hetero = False
+        for _, module in self.model.named_modules():
+            if isinstance(module, torch.nn.ModuleDict):
```
This might be a flaky way of detecting wether a model is heterogenous. Because a user might use a `ModuleDict` for other reasons too. 
Lets set `self.is_hetero` in `__call__` and then call `self.algorihm.support` again if `is_hetero` is `True`.

avgupta456 replied:
I think I understand what you're saying and pushed a commit, but not 100% sure on if the implementation is what you envisioned.

wsad1 commented on torch_geometric/explain/explainer.py:
```diff
@@ -140,22 +153,22 @@ def __call__(
-            explanation (Explanation): The explanation to threshold.
-        """
+    def _threshold_explanation(
+        self, explanation: Union[Explanation, HeteroExplanation]
+    ) -> Union[Explanation, HeteroExplanation]:
```
```suggestion
        self, explanation: Explanation
    ) -> Explanation:
```

wsad1 reviewed (commented):
Thanks. Left some initial comments.

wsad1 commented on torch_geometric/explain/algorithm/base.py:
```diff
@@ -20,12 +21,13 @@ class ExplainerAlgorithm(torch.nn.Module):
         model: torch.nn.Module,
-        x: Tensor,
-        edge_index: Tensor,
+        x: Union[Tensor, Dict[NodeType, Tensor]],
+        edge_index: Union[Tensor, Dict[EdgeType, Tensor]],
```
Update the docstring, to mention when `x` and `edge_index` need to be dicts.

wsad1 commented on torch_geometric/explain/algorithm/base.py:
```diff
@@ -43,14 +45,17 @@ def forward(
                 in a multi-task learning scenario. Should be kept to
                 :obj:`None` in case the model only returns a single output
                 tensor. (default: :obj:`None`)
+            is_hetero (bool, optional): Whether the input graph is
+                heterogeneous. (default: :obj:`False`)
```
We can detect if its Heterogenous if `x` is a dict we don't need the `is_hetero` arg.

wsad1 commented on torch_geometric/explain/explanation.py:
```diff
@@ -129,3 +131,136 @@ def _apply_masks(
+    feature-attribution. It can also hold the original graph if needed.
+
+    Args:
+        node_mask (Dict[NodeType, Tensor], optional): Dictionary of Node-level
+            mask with shape :obj:`[num_nodes]`. (default: :obj:`None`)
```
```suggestion
        node_mask (Dict[NodeType, Tensor], optional): Dictionary of Node-level
            mask with  key `NodeType` and value `Tensor` of 
            shape :obj:`[num_nodes]`. (default: :obj:`None`)
```
Nitpick: just to be clear that the shape of the dict isn't `num_nodes`.

avgupta456 commented:
I added logic and testing for thresholding, `validate()` function in heterogeneous explanation, and fixed some bugs. The last item left for this PR is creating heterogeneous explanation `subgraph`/`complement_subgraph` using `_apply_masks`. 

I am unsure if my implementation is good, overall having trouble extracting the `x_dict`, `edge_index_dict`, and `edge_attrs` from a `HeterogeneousExplanation`

Currently I am doing it like this

```python
x_dict = None
edge_index_dict = None
edge_attr_dict = None
for key, value in self.node_items():
    if key == "x_dict":
        x_dict = value
    elif key == "edge_index_dict":
        edge_index_dict = value
    elif key == "edge_attr_dict":
        edge_attr_dict = value
        continue
```

And setting the updated dictionaries like this:

```python
out.edge_index_dict.update(edge_index_dict)
out.edge_attr_dict.update(edge_attr_dict)
```

Is there a better way to access the `HeterogeneousExplanation` underlying graph?

wsad1 commented on torch_geometric/explain/explanation.py:
```diff
@@ -129,3 +131,188 @@ def _apply_masks(
+                edge_attr_dict = value
+
+        if edge_mask is not None:
+            for key, value in self.edge_mask.items():
+                edge_index_dict[key] = edge_index_dict[key][:, value]
+                edge_attr_dict[key] = edge_attr_dict[key][value]
```
```suggestion
            for edge_type in self.edge_types:
               out[edge_type].edge_index = out[edge_type].edge_index[:, value]
```
This might be cleaner. Also we need to check if `edge_attr` is in `out[edge_type]` before masking it.

wsad1 commented on torch_geometric/explain/algorithm/dummy_explainer.py:
```diff
@@ -46,5 +47,75 @@ def forward(
+
+        Args:
+            model (torch.nn.Module): The model to explain.
+            x (Union[torch.Tensor, Dict[NodeType, torch.Tensor]]): The node
+                features.
+            edge_index (Union[torch.Tensor, Dict[EdgeType, torch.Tensor]]):
+                The edge indices.
+            edge_attr (Union[torch.Tensor, Dict[EdgeType, torch.Tensor]],
+                optional): The edge attributes. (default: :obj:`None`)
```
Update doc string to indicate when `dict` is passed.

wsad1 commented on test/explain/algorithm/test_base.py:
```diff
@@ -0,0 +1,108 @@
+
+        explainer(data.x_dict, data.edge_index_dict)
+
+    explainer = Explainer(
+        model=model, algorithm=HeteroExplainerAlgorithm(),
+        explainer_config=ExplainerConfig(
+            explanation_type="model",
+            node_mask_type="object",
+            edge_mask_type=None,
+        ), model_config=ModelConfig(
+            mode="regression",
+            task_level="node",
+        ))
```
```suggestion
     explainer.algorithm = HeteroExplainerAlgorithm()
```

wsad1 commented on test/explain/algorithm/test_base.py:
```diff
@@ -0,0 +1,108 @@
+    data = hetero_data()
+    model = HeteroSAGE(data.metadata())
+
+    explainer = Explainer(
+        model=model, algorithm=AllExplainerAlgorithm(),
+        explainer_config=ExplainerConfig(
+            explanation_type="model",
+            node_mask_type="object",
+            edge_mask_type=None,
+        ), model_config=ModelConfig(
+            mode="regression",
+            task_level="node",
+        ))
+
+    explainer(data.x_dict, data.edge_index_dict)
```
Lets drop this test. it doesn't cover a different part of the code than the below two tests.

wsad1 commented on test/explain/test_explainer_hetero.py:
```diff
@@ -0,0 +1,204 @@
+from torch_geometric.nn import SAGEConv, to_hetero
+
+
+def get_edge_index(num_src_nodes, num_dst_nodes, num_edges):
+    row = torch.randint(num_src_nodes, (num_edges, ), dtype=torch.long)
+    col = torch.randint(num_dst_nodes, (num_edges, ), dtype=torch.long)
+    return torch.stack([row, col], dim=0)
+
+
+@pytest.fixture
+def data():
+    data = HeteroData()
+    data['paper'].x = torch.randn(8, 16)
+    data['author'].x = torch.randn(10, 8)
+    data['paper', 'paper'].edge_index = get_edge_index(8, 8, 10)
+    data['author', 'paper'].edge_index = get_edge_index(10, 8, 10)
+    data['paper', 'author'].edge_index = get_edge_index(8, 10, 10)
+    return data
+
+
+class GraphSAGE(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.conv1 = SAGEConv((-1, -1), 32)
+        self.conv2 = SAGEConv((-1, -1), 1)
+
+    def forward(self, x, edge_index):
+        x = self.conv1(x, edge_index).relu()
+        return self.conv2(x, edge_index)
+
+
+class HeteroSAGE(torch.nn.Module):
+    def __init__(self, metadata):
+        super().__init__()
+        self.graph_sage = to_hetero(GraphSAGE(), metadata, debug=False)
+
+    def forward(self, x_dict, edge_index_dict, *args) -> torch.Tensor:
+        return self.graph_sage(x_dict, edge_index_dict)['paper']
```
Lets move this to a `conftest.py` and share across all tests (https://docs.pytest.org/en/6.2.x/fixture.html#scope-sharing-fixtures-across-classes-modules-packages-or-session). Happy to discuss more about this or even address it in another PR.

avgupta456 replied:
Maybe best to move to a separate PR, I agree there is a lot of repeated code across the explanation tests that can be consolidated. Should I create an issue for this?

wsad1 replied:
Yeah feel free to add a TODO and also raise an issue.

avgupta456 replied:
Created #6186

wsad1 reviewed (commented):
Thanks for the updates @avgupta456 . Will take a look again once you are done with TODOs in test.

avgupta456 commented on torch_geometric/explain/explainer.py:
```diff
@@ -198,6 +200,72 @@ def _threshold(self, explanation: Explanation) -> Explanation:
+        self, mask_dict: Dict[str, Dict[Union[NodeType, EdgeType], Tensor]]
+    ) -> Dict[Union[NodeType, EdgeType], Tensor]:
+
+        if self.threshold_config.type == ThresholdType.hard:
+            mask_dict = {
+                key: {
+                    k: (v > self.threshold_config.value).float()
+                    for k, v in mask.items()
+                }
+                for key, mask in mask_dict.items()
+            }
+        elif self.threshold_config.type in [
+                ThresholdType.topk,
+                ThresholdType.topk_hard,
+        ]:
+            for key, mask in mask_dict.items():
+                for k, v in mask.items():
+                    if self.threshold_config.value >= v.numel():
+                        if self.threshold_config.type != ThresholdType.topk:
+                            mask_dict[key][k] = torch.ones_like(v)
+                        continue
+
+                    value, index = torch.topk(
+                        v.flatten(),
+                        k=self.threshold_config.value,
+                    )
+
+                    out = torch.zeros_like(v.flatten())
+                    if self.threshold_config.type == ThresholdType.topk:
+                        out[index] = value
+                    else:
+                        out[index] = 1.0
+                    mask_dict[key][k] = out.reshape(v.size())
+        else:
```
Yeah that's much simpler :)

avgupta456 commented:
I'm currently running into an issue with HeteroExplanation that is blocking my further progress.

Unlike Explanation (inherits from Data), HeteroExplanation (inherits from HeteroData) does not correctly store the x_dict, edge_index_dict, and edge_attrs_dict when initialized with these arguments. It does work if the dictionaries are set as attributes afterwards.

```python
HeteroExplanation(x_dict=x_dict, edge_index_dict=edge_index_dict, edge_attrs_dict=edge_attrs_dict)  # does not work

explanation = HeteroExplanation()  # does work
explanation.x_dict = x_dict
explanation.edge_index_dict = edge_index_dict
explanation.edge_attrs_dict = edge_attrs_dict
```

One workaround is that when initialized with the dictionaries, they are accessible through self.node_items(). Perhaps this is a bug with HeteroData? Or maybe I am doing something else incorrectly.

As a result of this bug, the subgraph method isn't working as intended. Open to any advice.

wsad1 commented on torch_geometric/explain/explanation.py:
```diff
@@ -129,3 +131,172 @@ def _apply_masks(
+            shape :obj:`[num_edges, num_edge_features]`. (default: :obj:`None`)
+        **kwargs (optional): Additional attributes.
+    """
+    def __init__(
+        self,
+        node_mask: Optional[Dict[NodeType, Tensor]] = None,
+        edge_mask: Optional[Dict[EdgeType, Tensor]] = None,
+        node_feat_mask: Optional[Dict[NodeType, Tensor]] = None,
+        edge_feat_mask: Optional[Dict[EdgeType, Tensor]] = None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+
+        # HeteroData doesn't set all kwargs as attributes, so we need to do it
```
```suggestion
    def __init__(
        x_dict, edge_index_dict, edge_attr_dict = None,
        node_mask: Optional[Dict[NodeType, Tensor]] = None,
        edge_mask: Optional[Dict[EdgeType, Tensor]] = None,
        node_feat_mask: Optional[Dict[NodeType, Tensor]] = None,
        edge_feat_mask: Optional[Dict[EdgeType, Tensor]] = None,
    ):
        super().__init__()
        for key, val in x_dict.items():
            self[key].x = val
        for key, val in edge_index_dict.items():
            self[key].edge_index = val
            if edge_attr_dict is not None:
                self.[key].edge_attr = edge_attr_dict[key]
        # HeteroData doesn't set all kwargs as attributes, so we need to do it
```
Assigning it this way should help you create a valid `HeteData` object on which `subgraph` will work. Note, I made the assumption that `x_dict` and `edge_index_dict` are passed as arguments, which is reasonable given that we create `HeteroExplanation` internally.

avgupta456 replied:
Thanks @wsad1! Resolving this comment as @dufourc1's comment is also below.

wsad1 commented on torch_geometric/explain/explanation.py:
```diff
@@ -129,3 +131,172 @@ def _apply_masks(
+
+        if edge_mask is not None:
+            for edge_type in self.edge_types:
+                if edge_type not in edge_mask:
+                    continue
```
Can we drop this? Because `validate` would fail if a `edge_type` didn't have a mask.

wsad1 reviewed (commented):
This looks clean now. Will approve once the `subgraph` issue is resolved.

dufourc1 reviewed (commented):
Thanks for the work ! LGTM, these are nitpick comments

wsad1 commented on torch_geometric/explain/explainer.py:
```diff
@@ -198,6 +200,40 @@ def _threshold(self, explanation: Explanation) -> Explanation:
+        """Threshold the explanation mask according to the thresholding method.
+
+        Args:
+            explanation (Explanation): The explanation to threshold.
```
```suggestion
            explanation (Explanation or HeteroExplanation): The explanation to threshold.
```

wsad1 commented on torch_geometric/explain/algorithm/dummy_explainer.py:
```diff
@@ -46,5 +47,74 @@ def forward(
+
+        Args:
+            model (torch.nn.Module): The model to explain.
+            x (Dict[NodeType, torch.Tensor]): The node features.
+            edge_index (Dict[EdgeType, torch.Tensor]): The edge indices.
+            edge_attr (Dict[EdgeType, torch.Tensor], optional): The edge
+                attributes. (default: :obj:`None`)
```
```suggestion
            x_dict (Dict[NodeType, torch.Tensor]): The node features for each node type.
            edge_index_dict (Dict[EdgeType, torch.Tensor]): The edge indices for each edge type.
            edge_attr_dict (Dict[EdgeType, torch.Tensor], optional): The edge
                attributes for each edge type. (default: :obj:`None`)
```

wsad1 commented on torch_geometric/data/hetero_data.py:
```diff
@@ -580,7 +580,7 @@ def subgraph(self, subset_dict: Dict[NodeType, Tensor]) -> 'HeteroData':
                 holding the nodes to keep for each node type.
         """
-        data = self.__class__(self._global_store)
+        data = self.__class__(**self._global_store)
```
Why do we need this change?

avgupta456 replied:
Since self.__class__ is HeteroExplanation, the dictionary self._global_store is interpreted as x_dict, and it complains for the other arguments (not to mention all the arguments are stored incorrectly in x_dict).

Since HeteroData() treats _mapping and kwargs together, this operation doesn't change the existing behavior and allows the arguments to be correctly mapped for HeteroExplanation.

wsad1 commented on test/explain/test_explainer_hetero.py:
```diff
@@ -0,0 +1,204 @@
+
+@pytest.mark.parametrize('threshold_value', list(range(1, 10)))
+@pytest.mark.parametrize('node_mask_type', ['object', 'attributes'])
+def test_topk_threshold(data, threshold_value, node_mask_type):
```
Can we combine topk and topk_hard into one test, using parametrization?

wsad1 commented on test/explain/algorithm/test_base.py:
```diff
@@ -0,0 +1,81 @@
+    data = hetero_data()
+    model = HeteroSAGE(data.metadata())
+
+    with pytest.raises(ValueError):
+        explainer = Explainer(
+            model=model, algorithm=HomoExplainerAlgorithm(),
+            explainer_config=ExplainerConfig(
+                explanation_type="model",
+                node_mask_type="object",
+                edge_mask_type=None,
+            ), model_config=ModelConfig(
+                mode="regression",
+                task_level="node",
+            ))
+
+        explainer(data.x_dict, data.edge_index_dict)
```
```suggestion

      explainer = Explainer(
          model=model, algorithm=HomoExplainerAlgorithm(),
          explainer_config=ExplainerConfig(
              explanation_type="model",
              node_mask_type="object",
              edge_mask_type=None,
          ), model_config=ModelConfig(
              mode="regression",
              task_level="node",
          ))
    with pytest.raises(ValueError):
        explainer(data.x_dict, data.edge_index_dict)
```

wsad1 reviewed (approved):
Thanks @avgupta456, this looks good.
